### PR TITLE
Refactored sympy.stats.frv_types, sympy.stats.frv and sympy.stats.drv_types

### PR DESCRIPTION
**Pull request by czgdp1807**
#### References to other Issues or PRs

N/A

#### Brief description of what is fixed or changed
The existing code of sympy.stats.frv_types is quite repeated at a lot of places in various distributions. In addition, the code of the above said file is not in alignment with the logic used in sympy.stats.crv_types and sympy.stats.drv_types for creating random variables of predefined distributions.
This PR aims at reducing the repeated and unwanted code and moving the logics and checks in `rv` function rather than defining checks and sympifying in the separate distribution classes.

#### Other comments
I was going through the code of sympy.stats.frv_types and hence, was prompted to make changes to enhance readability and consistency in the stats module.
Please **don't close** this PR. Instead, it would be great if you can discuss with me the various issues of merging this PR so that I can improve upon them.

Thanks for reading. Waiting for your comments. :-)

#### Release Notes

* stats
  * Refactored sympy.stats.frv_types and sympy.stats.frv

### CODE REVIEW

File: sympy/stats/frv_types.py [hunk 1]
```diff
@@ -16,34 +16,36 @@
...
 class FiniteDistributionHandmade(SingleFiniteDistribution):
     @property
     def dict(self):
         return self.args[0]
 
-    def __new__(cls, density):
-        density = Dict(density)
-        for k in density.values():
-            k_sym = sympify(k)
-            if fuzzy_not(fuzzy_and((k_sym.is_nonnegative, (k_sym - 1).is_nonpositive))):
```
Comment (smichr) [inline sympy/stats/frv_types.py]: looks like this was handling symbolic input. Does your modification allow the same?

Comment (czgdp1807) [reply sympy/stats/frv_types.py]: Yes. There tests for symbolic parameters which are passing, so I think my modifications doesn't cause any problem to those tests.

**Comment by czgdp1807** [comment]
@jksuom @Upabjojr Please review and let me know if any changes are to be made.

File: sympy/stats/frv_types.py [hunk 2]
```diff
@@ -174,14 +174,9 @@ def Die(name, sides=6):
...
-        p = args[BernoulliDistribution._argnames.index('p')]
-        p_sym = sympify(p)
-
-        if fuzzy_not(fuzzy_and((p_sym.is_nonnegative, (p_sym - 1).is_nonpositive))):
-            raise ValueError("p = %s is not in range [0, 1]." % str(p))
-        else:
-            return super(BernoulliDistribution, cls).__new__(cls, *args)
+    @staticmethod
+    def check(p, succ, fail):
+        _value_check(And(p >= 0, p <= 1), "p should be in range [0, 1].")
```
Comment (Upabjojr) [inline sympy/stats/frv_types.py]: careful, this may not be convertable into boolean.

Comment (czgdp1807) [reply sympy/stats/frv_types.py]: @Upabjojr This logic is used in drv_types too. An example will be helpful, so that changes can be made there as well.
Should I use `fuzzy_not` and `fuzzy_and`?

Comment (czgdp1807) [reply sympy/stats/frv_types.py]: This is the [line](https://github.com/sympy/sympy/blob/master/sympy/stats/drv_types.py#L46) I was referring too. Changes are to be made here too if `And(p >= 0, p <= 1)` is not convertible to boolean.

Comment (czgdp1807) [reply sympy/stats/frv_types.py]: @Upabjojr What are your views on the above the comments?

Comment (czgdp1807) [reply sympy/stats/frv_types.py]: @Upabjojr Any updates on this? Waiting for your comments.

Comment (czgdp1807) [reply sympy/stats/frv_types.py]: @Upabjojr I have updated the checks, kindly take a look and tell me if its fine.

File: sympy/stats/frv_types.py (hunk 1 above)
Comment (czgdp1807) [reply sympy/stats/frv_types.py]: @smichr Can I resolve this conversation?

File: sympy/stats/drv_types.py [hunk 3]
```diff
@@ -43,7 +43,8 @@ class GeometricDistribution(SingleDiscreteDistribution):
 
     @staticmethod
     def check(p):
-        _value_check(And(0 < p, p <= 1), "p must be between 0 and 1")
```
Comment (smichr) [inline sympy/stats/drv_types.py]: I had to look up `_value_check` because the action is not obvious from the name. How rename it to `_whenFalse(boolean, msg)`. And why did you remove the `And`? If either is False the `And` will be False.

File: sympy/stats/frv_types.py [hunk 4]
```diff
@@ -118,12 +121,10 @@ def DiscreteUniform(name, items):
...
 
-    def __new__(cls, sides):
-        sides_sym = sympify(sides)
-        if fuzzy_not(fuzzy_and((sides_sym.is_integer, sides_sym.is_positive))):
-            raise ValueError("'sides' must be a positive integer.")
-        else:
-            return super(DieDistribution, cls).__new__(cls, sides)
+    @staticmethod
+    def check(sides):
+        _value_check(sides.is_positive and sides.is_integer,
```
Comment (smichr) [inline sympy/stats/frv_types.py]: This needs the fuzzy logic or it will evaluate to False if one is unknown. The idiom has been to raise an error only if the condition is False, not if it is unevaluated.

File: sympy/stats/frv_types.py [hunk 5]
```diff
@@ -174,14 +175,10 @@ def Die(name, sides=6):
...
-        p = args[BernoulliDistribution._argnames.index('p')]
-        p_sym = sympify(p)
-
-        if fuzzy_not(fuzzy_and((p_sym.is_nonnegative, (p_sym - 1).is_nonpositive))):
-            raise ValueError("p = %s is not in range [0, 1]." % str(p))
-        else:
-            return super(BernoulliDistribution, cls).__new__(cls, *args)
+    @staticmethod
+    def check(p, succ, fail):
+        _value_check((p >= 0) != False and (p <= 1) != False,
```
Comment (smichr) [inline sympy/stats/frv_types.py]: ->`And(p >= 0, p <= 1)`

File: sympy/stats/frv_types.py [hunk 6]
```diff
@@ -259,18 +256,12 @@ def Coin(name, p=S.Half):
...
-
-        if fuzzy_not(fuzzy_and((n_sym.is_integer, n_sym.is_nonnegative))):
-            raise ValueError("'n' must be positive integer. n = %s." % str(n))
-        elif fuzzy_not(fuzzy_and((p_sym.is_nonnegative, (p_sym - 1).is_nonpositive))):
-            raise ValueError("'p' must be: 0 <= p <= 1 . p = %s" % str(p))
-        else:
-            return super(BinomialDistribution, cls).__new__(cls, *args)
+    @staticmethod
+    def check(n, p, succ, fail):
+        _value_check((n.is_integer and n.is_nonnegative),
```
Comment (smichr) [inline sympy/stats/frv_types.py]: `_whenFalse(fuzzy_and((n.is_integer, n.is_nonnegative)), "should be nonnegative integer: %s" % n)`

Comment (smichr) [reply sympy/stats/frv_types.py]: Actually, we are only interested in the case when this evaluates to False in an obvious way so building this with operators instead of assumptions is the way to go. I think the only thing extra needed to handle the integer assumption is
`is_integer = lambda x: Piecewise((Eq(floor(x), x),x>0),(Eq(-floor(-x),x), True))`. So this query could be
```python
_whenFalse(And(is_integer(n), n>=0), "expecting nonnegative integer, but got %s" % n)
```
There is already some built-in smarts to relationals:
```
>>> var('n', odd=True, positive=True, prime=True)
n
>>> n>0
True
>>> n>1
True
>>> n>2
True
>>> n>3
n > 3
```

File: sympy/stats/frv_types.py (hunk 4 above)
Comment (czgdp1807) [reply sympy/stats/frv_types.py]: i will update this asap.

File: sympy/stats/frv_types.py (hunk 5 above)
Comment (czgdp1807) [reply sympy/stats/frv_types.py]: I have removed `And` because of the [this](https://github.com/sympy/sympy/pull/16576#discussion_r282803523) discussion. What are your views?

File: sympy/stats/frv_types.py (hunk 6 above)
Comment (czgdp1807) [reply sympy/stats/frv_types.py]: i will make the changes may be by tomorrow. thanks for the reviews.

Comment (smichr) [reply sympy/stats/frv_types.py]: regarding previous...I think `_whenFalse(fuzzy_and([n.is_integer, n >= 0]), ...)` is fine. The rule is:
if `a` or `b` might be None, don't use `a and b` or And(a, b). Or, as you were doing (but in too many cases, I think), use `a != False` if `a` might be None. You don't have to do that if a is a Relational, however, because if a decision can't be made it remains a Relational (which is bool True) and won't confuse the `_whenFalse` tester.

File: sympy/stats/frv_types.py [hunk 7]
```diff
@@ -16,34 +16,37 @@
...
-                raise ValueError("Probability at a point must be between 0 and 1.")
-        sum_sym = sum(density.values())
-        if sum_sym != 1:
-            raise ValueError("Total Probability must be equal to 1.")
-        return Basic.__new__(cls, density)
+    @staticmethod
+    def check(density):
+        for p in density.values():
+            cond = (p >= 0) != False and (p <= 1) != False
+            _value_check(cond, "Probability at a point must be between 0 and 1.")
```
Comment (smichr) [inline sympy/stats/frv_types.py]: `And(p>=0, p<= 1)`

File: sympy/stats/frv_types.py (hunk 4 above)
Comment (smichr) [reply sympy/stats/frv_types.py]: `_whenFalse(fuzzy_and([sides.is_positive, sides.is_integer]), msg)`

**Comment by smichr** [comment]
I tried to do a PR to this but for some reason it pulled in many other commits. Here is a diff that can be used:
```diff
diff --git a/sympy/stats/joint_rv_types.py b/sympy/stats/joint_rv_types.py
index de45236..de1a0fe 100644
--- a/sympy/stats/joint_rv_types.py
+++ b/sympy/stats/joint_rv_types.py
@@ -80,8 +80,9 @@ def check(self, mu, sigma):
         _value_check(len(mu) == len(sigma.col(0)),
             "Size of the mean vector and covariance matrix are incorrect.")
         #check if covariance matrix is positive definite or not.
-        _value_check(all([Gt(i, 0) != False for i in sigma.eigenvals().keys()]),
-            "The covariance matrix must be positive definite. ")
+        all(_value_check(i > 0,
+            "The covariance matrix must be positive definite.")
+            for i in sigma.eigenvals().keys())
 
     def pdf(self, *args):
         mu, sigma = self.mu, self.sigma
@@ -121,8 +122,9 @@ def check(self, mu, sigma):
         _value_check(len(mu) == len(sigma.col(0)),
             "Size of the mean vector and covariance matrix are incorrect.")
         #check if covariance matrix is positive definite or not.
-        _value_check(all([Gt(i, 0) != False for i in sigma.eigenvals().keys()]),
-            "The covariance matrix must be positive definite. ")
+        all(_value_check(i > 0,
+                    "The covariance matrix must be positive definite.")
+                    for i in sigma.eigenvals().keys())
 
     def pdf(self, *args):
         mu, sigma = self.mu, self.sigma
@@ -155,8 +157,9 @@ def check(self, mu, sigma, v):
         _value_check(len(mu) == len(sigma.col(0)),
             "Size of the location vector and shape matrix are incorrect.")
         #check if covariance matrix is positive definite or not.
-        _value_check(all([Gt(i, 0) != False for i in sigma.eigenvals().keys()]),
-            "The shape matrix must be positive definite. ")
+        all(_value_check(i > 0,
+                    "The shape matrix must be positive definite.")
+                    for i in sigma.eigenvals().keys())
 
     def pdf(self, *args):
         mu, sigma = self.mu, self.shape_mat
@@ -260,9 +263,9 @@ class MultivariateBetaDistribution(JointDistribution):
 
     def check(self, alpha):
         _value_check(len(alpha) >= 2, "At least two categories should be passed.")
-        for a_k in alpha:
-            _value_check((a_k > 0) != False, "Each concentration parameter"
-                                            " should be positive.")
+        all(_value_check(a_k > 0,
+            "Each concentration paramenter should be positive.")
+            for a_k in alpha)
 
     @property
     def set(self):
@@ -332,7 +335,7 @@ class MultivariateEwensDistribution(JointDistribution):
     is_Continuous = False
 
     def check(self, n, theta):
-        _value_check(isinstance(n, Integer) and (n > 0) == True,
+        _value_check(And(isinstance(n, Integer), n > 0),
                         "sample size should be positive integer.")
         _value_check(theta.is_positive, "mutation rate should be positive.")
 
@@ -404,10 +407,10 @@ class MultinomialDistribution(JointDistribution):
     is_Discrete = True
 
     def check(self, n, p):
-        _value_check(((n > 0) != False) and isinstance(n, Integer),
+        _value_check(And(n > 0, isinstance(n, Integer)),
                         "number of trials must be a positve integer")
         for p_k in p:
-            _value_check((p_k >= 0) != False,
+            _value_check(p_k >= 0,
                         "probability must be at least a positive symbol.")
 
     @property
@@ -471,10 +474,10 @@ class NegativeMultinomialDistribution(JointDistribution):
     is_Discrete = True
 
     def check(self, k0, p):
-        _value_check(((k0 > 0) != False) and isinstance(k0, Integer),
+        _value_check(And(k0 > 0, isinstance(k0, Integer)),
                         "number of failures must be a positve integer")
         for p_k in p:
-            _value_check((p_k >= 0) != False,
+            _value_check(p_k >= 0,
                         "probability must be at least a positive symbol.")
 
     @property
diff --git a/sympy/stats/rv.py b/sympy/stats/rv.py
index 3b44457..2feb4c2 100644
--- a/sympy/stats/rv.py
+++ b/sympy/stats/rv.py
@@ -1318,9 +1318,36 @@ def __getattr__(self, attr):
 
 def _value_check(condition, message):
     """
-    Check a condition on input value.
+    Raise a ValueError with message if condition is False, else
+    return True.
 
-    Raises ValueError with message if condition is not True
+    Examples
+    ========
+
+    >>> from sympy.stats.rv import _value_check
+    >>> from sympy.abc import a, b, c
+    >>> from sympy import And, Dummy
+
+    >>> _value_check(2 < 3, '')
+    True
+    >>> _value_check(a < b, '')
+    True
+    >>> r = Dummy(real=True)
+    >>> _value_check(r < r - 1, 'msg')
+    Traceback (most recent call last):
+    ...
+    ValueError: msg
+
+    If many conditions must not be False, they can be checked
+    with all or And. Each of the following is equivalent:
+
+    >>> all(_value_check(i < 0, '') for i in (a, b, c))
+    True
+    >>> _value_check(And(*[i < 0 for i in (a, b, c)]), '')
+    True
+    >>> _value_check(And(a < 0, b < 0, c < 0), '')
+    True
     """
     if condition == False:
         raise ValueError(message)
+    return True
```

**Comment by czgdp1807** [comment]
Thank you @smichr for the reviews and the diff. I will start working on PRs from around 15:00 IST to 20:30 IST and will update this one too in the process.

File: sympy/stats/drv_types.py (hunk 3 above)
Comment (czgdp1807) [reply sympy/stats/drv_types.py]: We can import `_value_check` as `_whenFalse` wherever necessary. Renaming can cause a lot of efforts to be made to make changes across the stats module. I think it will be less useful and after all this depicts more of a private function due to `_` in front of it  and hence, only developers have to understand it's purpose which I believe is we can interpret from its few lines of code in `sympy.stats.rv`. What do you say?

File: sympy/stats/frv_types.py (hunk 6 above)
Comment (czgdp1807) [reply sympy/stats/frv_types.py]: I don't think Relational is bool true, I am getting the following error,
```
Python 3.6.7 (default, Oct 22 2018, 11:32:17) 
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> var('n', odd=True, positive=True, prime=True)
n
>>> n>0
True
>>> n>1
True
>>> n>2
True
>>> n>3
n > 3
>>> n>3 == True
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/gagandeep/sympy/sympy/core/relational.py", line 305, in __nonzero__
    raise TypeError("cannot determine truth value of Relational")
TypeError: cannot determine truth value of Relational

```

Comment (czgdp1807) [reply sympy/stats/frv_types.py]: Can we apply the below mentioned principles:

Let's say there are two conditions, `cond1` and `cond2`, the following should be done,

1. If `cond1` or `cond2` might be `None`, then use `fuzzy_and` and `fuzzy_not`.
2. If `cond1` and `cond2` are both determined to be either `True` or `False` and nothing else then use `And` or `and`.

What do you say @smichr? Can I go with the above rules during my summer project?

Comment (jksuom) [reply sympy/stats/frv_types.py]: `n > 3` should be parenthesized:
```
>>> n = symbols('n')
>>> (n > 3) == True
False
```

Comment (czgdp1807) [reply sympy/stats/frv_types.py]: So is Relational, bool True, or bool False or None?

Comment (jksuom) [reply sympy/stats/frv_types.py]: Or none of those (in which case it is not equal to True). If you write `n > 3 == True`, Python will interpret it as `n > 3 and 3 == True` and therefore needs to evaluate `n > 3` as a boolean.

Comment (czgdp1807) [reply sympy/stats/frv_types.py]: So what to use, in such cases. `And` or `fuzzy_and`?

Comment (jksuom) [reply sympy/stats/frv_types.py]: Perhaps it would be possible to use two checks with different error messages: `_value_check(n.is_integer, "... expecting integer...")` and similarly for `n.is_nonnegative`.

Comment (smichr) [reply sympy/stats/frv_types.py]: Please clarify the specific case you are referring to.

And note: `is_foo` attributes will be True, False or None while Relational will be True, False or Relational.

Comment (czgdp1807) [reply sympy/stats/frv_types.py]: I think that's the best way, consider the current situation. I will re-write the checks for this PR by avoiding `And`, `fuzzy_and`. Thanks for helping me out with this confusion. @smichr @Upabjojr I am going with this suggestion.

Comment (czgdp1807) [reply sympy/stats/frv_types.py]: > Please clarify the specific case you are referring to.

Like, if p > 0 and p < 1 are undetermined, I mean p can be symbols, so what to use in such cases, `And` or `fuzzy_and`?

Comment (smichr) [reply sympy/stats/frv_types.py]: Or, as in the diff: 
```python
_value_check(And(n>=0, n.is_Integer), 'expecting literal nonnegative integer')
_value_check(fuzzy_and([n>=0, n.is_integer]), 'expecting nonnegative integer')
```
We could dispence with any sort of wrapper by making the `_value_check signature be 
```
def _value_check(cond, msg):
    if not iterable(cond):
        cond = [cond]
    if fuzzy_and(cond) == False:
            raise ValueError(msg)
    return True
```

File: sympy/stats/drv_types.py [hunk 8]
```diff
@@ -19,6 +19,7 @@
 from sympy.stats.drv import SingleDiscreteDistribution, SingleDiscretePSpace
 from sympy.stats.joint_rv import JointPSpace, CompoundDistribution
 from sympy.stats.rv import _value_check, RandomSymbol
+from sympy.core.logic import fuzzy_and, fuzzy_not
```
Comment (smichr) [inline sympy/stats/drv_types.py]: unused

**Comment by czgdp1807** [comment]
@smichr @jksuom I have updated the PR after considering the following points,

1. For conditions like, p > 0, I have used, `(p > 0) != False`.
2. For conditions, like, p > 0 and p < 1, I have used `And(p > 0, p < 1)`
3. For conditions, like, p.is_integer and p.is_positive, I have used, `fuzzy_and([p.is_integer, p.is_positive])`

I arrived at the above points after reading the whole discussion again happened over various PRs opened by me recently. Please comment your suggestions if there is anything extra to be done. I was confused a bit but I have tried my best to accommodate everyone's suggestions.
Thanks for continuously helping me with the concepts and quick responses. :-)

File: sympy/stats/frv_types.py (hunk 6 above)
Comment (czgdp1807) [reply sympy/stats/frv_types.py]: @smichr I think we can work on `_value_check` on a different PR in future because, according to me the function currently doesn't take many cases into account and few them you have suggested in your comment.

**Comment by czgdp1807** [comment]
This PR only aims to refactor `frv`,  `drv_types` and `frv_types`. I will refactor `joint_rv_types` later, may be at the end of the Bonding Period as I am currently facing issues with a higher priority in that file.

**Comment by smichr** [comment]
```diff
diff --git a/sympy/stats/drv_types.py b/sympy/stats/drv_types.py
index d5a2c7b..1615a06 100644
--- a/sympy/stats/drv_types.py
+++ b/sympy/stats/drv_types.py
@@ -13,7 +13,7 @@
 
 from __future__ import print_function, division
 
-from sympy import (factorial, exp, S, sympify, And, I, zeta, polylog, log, beta, hyper, binomial,
+from sympy import (factorial, exp, S, sympify, I, zeta, polylog, log, beta, hyper, binomial,
                    Piecewise, floor)
 from sympy.stats import density
 from sympy.stats.drv import SingleDiscreteDistribution, SingleDiscretePSpace
@@ -43,7 +43,7 @@ class GeometricDistribution(SingleDiscreteDistribution):
 
     @staticmethod
     def check(p):
-        _value_check(And(0 < p, p <= 1), "p must be between 0 and 1")
+        _value_check((0 < p, p <= 1), "p must be between 0 and 1")
 
     def pdf(self, k):
         return (1 - self.p)**(k - 1) * self.p
@@ -115,7 +115,7 @@ class LogarithmicDistribution(SingleDiscreteDistribution):
 
     @staticmethod
     def check(p):
-        _value_check(And(p > 0, p < 1), "p should be between 0 and 1")
+        _value_check((p > 0, p < 1), "p should be between 0 and 1")
 
     def pdf(self, k):
         p = self.p
@@ -192,8 +192,8 @@ class NegativeBinomialDistribution(SingleDiscreteDistribution):
 
     @staticmethod
     def check(r, p):
-        _value_check((r > 0) != False, 'r should be positive')
-        _value_check(And(p > 0, p < 1), 'p should be between 0 and 1')
+        _value_check(r > 0, 'r should be positive')
+        _value_check((p > 0, p < 1), 'p should be between 0 and 1')
 
     def pdf(self, k):
         r = self.r
@@ -279,7 +279,7 @@ class PoissonDistribution(SingleDiscreteDistribution):
 
     @staticmethod
     def check(lamda):
-        _value_check((lamda > 0) != False, "Lambda must be positive")
+        _value_check(lamda > 0, "Lambda must be positive")
 
     def pdf(self, k):
         return self.lamda**k / factorial(k) * exp(-self.lamda)
@@ -369,7 +369,7 @@ class YuleSimonDistribution(SingleDiscreteDistribution):
 
     @staticmethod
     def check(rho):
-        _value_check((rho > 0) != False, 'rho should be positive')
+        _value_check(rho > 0, 'rho should be positive')
 
     def pdf(self, k):
         rho = self.rho
@@ -448,7 +448,7 @@ class ZetaDistribution(SingleDiscreteDistribution):
 
     @staticmethod
     def check(s):
-        _value_check((s > 1) != False, 's should be greater than 1')
+        _value_check(s > 1, 's should be greater than 1')
 
     def pdf(self, k):
         s = self.s
diff --git a/sympy/stats/frv_types.py b/sympy/stats/frv_types.py
index 75915ad..c6e3344 100644
--- a/sympy/stats/frv_types.py
+++ b/sympy/stats/frv_types.py
@@ -16,11 +16,10 @@
 from __future__ import print_function, division
 
 from sympy import (S, sympify, Rational, binomial, cacheit, Integer,
-        Dict, Basic, KroneckerDelta, Dummy, And, Eq)
+        Dict, Basic, KroneckerDelta, Dummy, Eq)
 from sympy.concrete.summations import Sum
 from sympy.core.compatibility import as_int, range
 from sympy.stats.rv import _value_check
-from sympy.core.logic import fuzzy_and, fuzzy_not
 from sympy.stats.frv import (SingleFinitePSpace, SingleFiniteDistribution)
 
 __all__ = ['FiniteRV', 'DiscreteUniform', 'Die', 'Bernoulli', 'Coin',
@@ -45,7 +44,7 @@ def dict(self):
     @staticmethod
     def check(density):
         for p in density.values():
-            _value_check(And(p >= 0, p <= 1),
+            _value_check((p >= 0, p <= 1),
                         "Probability at a point must be between 0 and 1.")
         _value_check(Eq(sum(density.values()), 1), "Total Probability must be 1.")
 
@@ -124,7 +123,7 @@ class DieDistribution(SingleFiniteDistribution):
 
     @staticmethod
     def check(sides):
-        _value_check(fuzzy_and([sides.is_positive, sides.is_integer]),
+        _value_check((sides.is_positive, sides.is_integer),
                     "number of sides must be a positive integer.")
 
     @property
@@ -178,7 +177,7 @@ class BernoulliDistribution(SingleFiniteDistribution):
 
     @staticmethod
     def check(p, succ, fail):
-        _value_check(And(p >= 0, p <= 1),
+        _value_check((p >= 0, p <= 1),
                     "p should be in range [0, 1].")
 
     @property
@@ -259,9 +258,9 @@ class BinomialDistribution(SingleFiniteDistribution):
 
     @staticmethod
     def check(n, p, succ, fail):
-        _value_check(fuzzy_and([n.is_integer, n.is_nonnegative]),
+        _value_check((n.is_integer, n.is_nonnegative),
                     "'n' must be nonnegative integer.")
-        _value_check((p <= 1) != False and (p >= 0) != False,
+        _value_check((p <= 1, p >= 0),
                     "p should be in range [0, 1].")
 
     @property
diff --git a/sympy/stats/joint_rv_types.py b/sympy/stats/joint_rv_types.py
index b973729..7b91f6e 100644
--- a/sympy/stats/joint_rv_types.py
+++ b/sympy/stats/joint_rv_types.py
@@ -1,4 +1,4 @@
-from sympy import (sympify, S, pi, sqrt, exp, Lambda, Indexed, Gt,
+from sympy import (sympify, S, pi, sqrt, exp, Lambda, Indexed,
     IndexedBase)
 from sympy.matrices import ImmutableMatrix
 from sympy.matrices.expressions.determinant import det
@@ -79,7 +79,7 @@ def check(self, mu, sigma):
         _value_check(len(mu) == len(sigma.col(0)),
             "Size of the mean vector and covariance matrix are incorrect.")
         #check if covariance matrix is positive definite or not.
-        _value_check(all([Gt(i, 0) != False for i in sigma.eigenvals().keys()]),
+        _value_check((i > 0 for i in sigma.eigenvals().keys()),
             "The covariance matrix must be positive definite. ")
 
     def pdf(self, *args):
@@ -120,7 +120,7 @@ def check(self, mu, sigma):
         _value_check(len(mu) == len(sigma.col(0)),
             "Size of the mean vector and covariance matrix are incorrect.")
         #check if covariance matrix is positive definite or not.
-        _value_check(all([Gt(i, 0) != False for i in sigma.eigenvals().keys()]),
+        _value_check((i > 0 for i in sigma.eigenvals().keys()),
             "The covariance matrix must be positive definite. ")
 
     def pdf(self, *args):
@@ -155,7 +155,7 @@ def check(self, mu, sigma, v):
         _value_check(len(mu) == len(sigma.col(0)),
             "Size of the location vector and shape matrix are incorrect.")
         #check if covariance matrix is positive definite or not.
-        _value_check(all([Gt(i, 0) != False for i in sigma.eigenvals().keys()]),
+        _value_check((i > 0 for i in sigma.eigenvals().keys()),
             "The shape matrix must be positive definite. ")
 
     def pdf(self, *args):
diff --git a/sympy/stats/rv.py b/sympy/stats/rv.py
index 413dbb0..756b6f2 100644
--- a/sympy/stats/rv.py
+++ b/sympy/stats/rv.py
@@ -1273,9 +1273,59 @@ def __getattr__(self, attr):
 
 def _value_check(condition, message):
     """
-    Check a condition on input value.
+    Raise a ValueError with message if condition is False, else
+    return True if all conditions were True, else False.
 
-    Raises ValueError with message if condition is not True
+    Examples
+    ========
+
+    >>> from sympy.stats.rv import _value_check
+    >>> from sympy.abc import a, b, c
+    >>> from sympy import And, Dummy
+
+    >>> _value_check(2 < 3, '')
+    True
+
+    Here, the condition is not False, but it doesn't evaluate to True
+    so False is returned (but no error is raised). So checking if the
+    return value is True or False will tell you if all conditions were
+    evaluated.
+
+    >>> _value_check(a < b, '')
+    False
+
+    In this case the condition is False so an error is raised:
+
+    >>> r = Dummy(real=True)
+    >>> _value_check(r < r - 1, 'condition is not true')
+    Traceback (most recent call last):
+    ...
+    ValueError: condition is not true
+
+    If no condition of many conditions must be False, they can be
+    checked by passing them as an iterable:
+
+    >>> _value_check((a < 0, b < 0, c < 0), '')
+    False
+
+    The iterable can be a generator, too:
+
+    >>> _value_check((i < 0 for i in (a, b, c)), '')
+    False
+
+    The following are equivalent to the above but do not pass
+    an iterable:
+
+    >>> all(_value_check(i < 0, '') for i in (a, b, c))
+    False
+    >>> _value_check(And(a < 0, b < 0, c < 0), '')
+    False
     """
-    if condition == False:
+    from sympy.core.compatibility import iterable
+    from sympy.core.logic import fuzzy_and
+    if not iterable(condition):
+        condition = [condition]
+    truth = fuzzy_and(condition)
+    if truth == False:
         raise ValueError(message)
+    return truth == True
```

This diff passes the stats tests.

**Comment by smichr** [comment]
It would be nice to apply that diff and then squash the last 10 commits into the first to clean up the commit history.

**Comment by czgdp1807** [comment]
Sure I will do it. I hope other members involved here will agree on the diff you provided. I will do the update by tomorrow night so that, others have enough time to comment.

**Comment by smichr** [comment]
> other members involved here will agree on the diff

Just to be clear, the only functionality that I have changed is to allow `_value_check` to check more than one value. This factors out the need to use `!= False` on non-Relational tests and the need to use `And` or `fuzzy_and` on more than one test. Now `_value_check` will raise an error if any arg is False, will return False if any arg is not True, else will return True.

File: sympy/stats/drv_types.py (hunk 3 above)
Comment (sidhantnagpal) [reply sympy/stats/drv_types.py]: > I had to look up `_value_check` because the action is not obvious from the name. How rename it to `_whenFalse(boolean, msg)`.

I agree, I had to do the same. `_whenFalse` would be a better name.

**Comment by czgdp1807** [comment]
ping @smichr
